desci-server/src/controllers/nodes/publish.ts

```diff
@@ -98,49 +102,270 @@ export const publish = async (req: PublishRequest, res: Response<PublishResBody>
+    where: {
+      id: -1,
+      nodeId: node.id,
+    },
+    orderBy: {
+      id: "desc",
+    },
+  });
+
+  // Prevent duplicating the NodeVersion entry if the latest version is the same as the one we're trying to publish, as a draft save is triggered before publishing
+  const latestNodeVersionId = latestNodeVersion?.manifestUrl === cid
+    ? latestNodeVersion.id
+    : -1;
+
+  const nodeVersion = await prisma.nodeVersion.upsert({
+    where: {
+      id: latestNodeVersionId,
+    },
+    update: {
+      commitId,
+    },
+    create: {
+      nodeId: node.id,
+      manifestUrl: cid,
+      commitId,
+    },
+  });
+
+  // first time we see a stream for this node, make sure we bind it in the db
+  if (!node.ceramicStream) {
+    logger.trace(`[publish:publish] setting streamID ${ceramicStream} on node ${uuid}`);
+    await setCeramicStream(uuid, ceramicStream);
+  } else if (node.ceramicStream !== ceramicStream) {
+    logger.warn(
+      // This is unexpected and weird, but important to know if it occurs
+      `[publish:publish] stream on record does not match passed streamID`,
+      { database: node.ceramicStream, ceramicStream },
+    );
+  };
+
+  const legacyDpid = manifest.dpid?.id ? parseInt(manifest.dpid.id) : undefined;
+  let dpidAlias: number = node.dpidAlias;
+
+  // Do dataRef and dPID registration operations concurrently
+  const promises = [];
+
+  if (!dpidAlias) {
+    // The only reason this isn't just fire-and-forget is that we want the dpid
+    // for the discord notification, which won't be available otherwise for
+    // first time publishes.
+    promises.push(
+      createOrUpgradeDpidAlias(legacyDpid, ceramicStream, uuid)
+      .then(dpid => dpidAlias = dpid)
+    );
+  };
+
+  promises.push(
+    handlePublicDataRefs({
+      nodeId: node.id,
+      userId: owner.id,
+      manifestCid: cid,
+      nodeVersionId: nodeVersion.id,
+      nodeUuid: node.uuid,
+    })
+  );
+
+  await Promise.all(promises);
+
+  // TODO: different resolver url for codex? :thinking:
+  const targetDpidUrl = getTargetDpidUrl();
+
+  discordNotify(`${targetDpidUrl}/${dpidAlias}`);
+
+  /**
+   * Save the cover art for this Node for later sharing: PDF -> JPG for this version
+   */
+  cacheNodeMetadata(node.uuid, cid);
+};
+
+/**
+ * Creates new dPID if legacyDpid is falsy, otherwise tries to upgrade
+ * the dPID by binding the stream in the alias registry for that dPID.
+*/
+const createOrUpgradeDpidAlias = async (
+  legacyDpid: number | undefined,
+  ceramicStream: string,
+  uuid: string,
+): Promise<number> => {
+  let dpidAlias: number;
+  if (legacyDpid) {
+    // Requires the REGISTRY_OWNER_PKEY to be set in env
+    dpidAlias = await upgradeDpid(legacyDpid, ceramicStream);
+  } else {
+    // Will nicely return the existing dpid if this is called multiple times
+    dpidAlias = await getOrCreateDpid(ceramicStream);
+  };
+  await setDpidAlias(uuid, dpidAlias);
+  return dpidAlias;
```

Comment:
Depending on whether the manifest holds a legacy dPID entry, we decide to either claim the original dPID entry or create a brand new dPID